rs/replicated_state/src/canister_state/system_state/log_memory_store/struct_io.rs

```diff
@@ -0,0 +1,458 @@
+    }
+
+    pub fn load_header(&self) -> Header {
+        let (magic, addr) = self.read_raw_bytes::<3>(HEADER_OFFSET);
+        let (version, addr) = self.read_raw_u8(addr);
+        let (index_table_pages, addr) = self.read_raw_u16(addr);
+        let (index_entries_count, addr) = self.read_raw_u16(addr);
+        let (data_offset, addr) = self.read_raw_u64(addr);
+        let (data_capacity, addr) = self.read_raw_u64(addr);
+        let (data_size, addr) = self.read_raw_u64(addr);
+        let (data_head, addr) = self.read_raw_u64(addr);
+        let (data_tail, addr) = self.read_raw_u64(addr);
+        let (next_idx, addr) = self.read_raw_u64(addr);
+        let (max_timestamp, _addr) = self.read_raw_u64(addr);
+        Header {
+            magic,
+            version,
+            index_table_pages,
+            index_entries_count,
+            data_offset: MemoryAddress::new(data_offset),
+            data_capacity: MemorySize::new(data_capacity),
+            data_size: MemorySize::new(data_size),
+            data_head: MemoryPosition::new(data_head),
+            data_tail: MemoryPosition::new(data_tail),
+            next_idx,
+            max_timestamp,
+        }
+    }
+
+    pub fn save_header(&mut self, header: &Header) {
+        let mut addr = HEADER_OFFSET;
+        addr = self.write_raw_bytes(addr, &header.magic);
+        addr = self.write_raw_u8(addr, header.version);
+        addr = self.write_raw_u16(addr, header.index_table_pages);
+        addr = self.write_raw_u16(addr, header.index_entries_count);
+        addr = self.write_raw_u64(addr, header.data_offset.get());
+        addr = self.write_raw_u64(addr, header.data_capacity.get());
+        addr = self.write_raw_u64(addr, header.data_size.get());
+        addr = self.write_raw_u64(addr, header.data_head.get());
+        addr = self.write_raw_u64(addr, header.data_tail.get());
+        addr = self.write_raw_u64(addr, header.next_idx);
+        _ = self.write_raw_u64(addr, header.max_timestamp);
+    }
+
+    pub fn load_index_table(&self) -> IndexTable {
+        let h = self.load_header();
+        let pos = h.data_head;
+        let front = self
+            .load_record_without_content(pos)
+            .map(|record| IndexEntry::new(pos, &record));
+        let entries = if h.index_entries_count == 0 {
+            vec![]
+        } else {
+            let mut entries = Vec::with_capacity(h.index_entries_count as usize);
+            let mut addr = INDEX_TABLE_OFFSET;
+            for _ in 0..h.index_entries_count {
+                let (entry, next_addr) = self.read_index_entry(addr);
+                entries.push(entry);
+                addr = next_addr;
+            }
+            entries
+        };
+        IndexTable::new(
+            front,
+            h.data_capacity,
+            h.index_table_pages,
+            RESULT_MAX_SIZE,
+            entries,
+        )
+    }
+
+    pub fn save_index_table(&mut self, index: &IndexTable) {
```

Comment:
This requires that the header has already been saved. I guess there are other cases where the header needs to already have been saved as well - should we maybe be automatically loading a header or saving the default header when creating a new StructIo to avoid this requirement? Otherwise they should maybe be documented.